core/src/main/java/io/grpc/ClientInterceptors.java

```diff
@@ -204,4 +216,190 @@ public final void start(Listener<RespT> responseListener, Metadata headers) {
       }
     }
   }
+
+  public static class RetryingInterceptor implements ClientInterceptor {
+    // TODO(lukaszx0) have default logger for this one to give more visibility into when/what gets
+    // retried?
+    // private final Logger logger = Logger.getLogger(RetryingInterceptor.class.getName());
+    private final RetryPolicy retryPolicy;
+
+    public RetryingInterceptor(RetryPolicy retryPolicy) {
+      this.retryPolicy = retryPolicy;
+    }
+
+    public RetryingInterceptor() {
+      this.retryPolicy = new IdempotentRetryPolicy();
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+            MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+      return new RetryingCall<ReqT, RespT>(retryPolicy, method, callOptions, next);
+    }
+
+    private class RetryingCall<ReqT, RespT> extends ClientCall<ReqT, RespT> {
+      private final RetryPolicy retryPolicy;
+      private final MethodDescriptor<ReqT, RespT> method;
+      private final CallOptions callOptions;
+      private final Channel channel;
+      private final ScheduledExecutorService scheduledExecutor;
+      private Listener<RespT> responseListener;
+      private Metadata requestHeaders;
+      private ReqT requestMessage;
+      private boolean compressionEnabled;
+      private final Queue<AttemptListener> attemptListeners =
+              new ConcurrentLinkedQueue<AttemptListener>();
+      private volatile AttemptListener latestResponse;
+      private volatile ScheduledFuture<?> retryTask;
+
+      RetryingCall(RetryPolicy retryPolicy, MethodDescriptor<ReqT, RespT> method,
+                   CallOptions callOptions, Channel channel) {
+        this.method = method;
+        this.callOptions = callOptions;
+        this.channel = channel;
+        this.retryPolicy = retryPolicy;
+        this.scheduledExecutor = SharedResourceHolder.get(TIMER_SERVICE);
+      }
+
+      @Override
+      public void start(Listener<RespT> listener, Metadata headers) {
+        checkState(attemptListeners.isEmpty());
+        checkState(responseListener == null);
+        checkState(requestHeaders == null);
+        responseListener = listener;
+        requestHeaders = headers;
+        ClientCall<ReqT, RespT> firstCall = channel.newCall(method, callOptions);
+        AttemptListener attemptListener = new AttemptListener(firstCall);
+        attemptListeners.add(attemptListener);
+        firstCall.start(attemptListener, headers);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        lastCall().request(numMessages);
+      }
+
+      @Override
+      public void cancel() {
+        for (AttemptListener attempt : attemptListeners) {
```

Comment:
It looks like we have a race condition here: between one thread cancelling the call (and this method propagating to all outstanding attempts) and a scheduled retry creating a new attempt. It is possible for the retry to be added to the attempt queue *after* the call has been cancelled -- and thus we leave that attempt outstanding instead of also cancelling it.

This can be solved with a volatile flag to track if cancel has been called. When initiating a call (in both `start` and `maybeRetry`), we need to check the flag and skip the call if we're already cancelled. To avoid the race condition we need check the flag *again* after adding the attempt to the queue. If we're suddenly cancelled, go ahead and cancel the attempt just added.

Also -- right now, we really only need to keep track of the last call since we schedule them serially. In other words, if there is more than one attempt in the queue, only the last one will be outstanding and all others will already be finished. But if we intend to implement retry timeouts (where we may have more than one attempt concurrently outstanding), then we will need to keep this structure. Maybe a TODO about this?

For now, you can keep the queue, but remove attempts from it when they complete. (So the queue will in practice only have one element, until retry timeout support is added.)